load next card in background

Arthur-Milchior opened the pull request:
Sometime, loading the next question takes more than a second. This PR simply precompute the next question (finding the card from the database, computing question and answer side), so that the next question can be loaded quickly. The next card is discarded as soon as there is a reason to do so (e.g. when reset() is done, when it's buried by a sibling.) 

This is of course done in background.

This was initially a part of the bug PR "reset in background", and then I relazied that it was actually a self-contained commit.

Note that the constant is 52 because I used constant 50 and 51 in other branch/PR. It's easier for me to keep this number, this will ensure that I have less merge problem when multiple PR are accepted. Historically, it seems it did happens quite often in the past weeks, since I concentrate on one part of the code specially.

I want to emphasize that, with this PR, even with complex card type, I see no lag at all between two cards, and that's extremely nice. Alas the time is still long enough for the buttons to flicker. 


I ensured that the difference with upstream anki is as small as possible. I could of course pre load the value in the reviewer directly, so that there is literally no difference in the scheduler, but I believe that this would make the code less clear. In particular because the nice thing with the code as it is right now is that, even if the card is removed from it's queue, the count remains correct. This would not be the case if the scheduler had really sent the card to the reviewer

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/AbstractSched.java:
```diff
@@ -232,4 +232,10 @@ public String toString() {
         int operation(long did, int lim);
     }
+
+    // Should be called in background, preload next question/answer.
+    // Synchronzide, so that if the nextCard is already computed, it
+    // does not get overridden.
+    // Not in libanki. In AnkiDroid to save time
```
```suggestion
    // Preload next question/answer.
    // Not in libanki. In AnkiDroid to improve interactive responsiveness
```

Just comment tweaks for precision -

- the synchronization is an implementation detail and it is not actually synchronized here in the definition
- it actually spends the same amount of application time, possibly more if pre-computation is thrown away so it is not really saving time, but it's not useless at all - it improves responsiveness to the user a lot

Arthur-Milchior replied:
I wrote the comment thinking of myself as a user. As a user who hates waiting even a second for the next question. And it does save me time.
I never thought for a second that it was saving computation time; however that's not something I cared about here

I will however admit that my writting may be ambiguous, especially since I've been known to reduce computation time of a lot of task in those past two weeks.

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.java:
```diff
@@ -1460,6 +1460,10 @@ protected void _burySiblings(Card card) {
                     }
                     mNewQueue.remove(cid);
                 }
+                if (mNextCard != null && mNextCard.getId() == cid && ((buryRev && mNextCard.getQueue() == Consts.QUEUE_TYPE_REV)) || ((buryNew && mNextCard.getQueue() == Consts.QUEUE_TYPE_NEW))) {
```
This is pretty involved! Could you put a comment on top of this line that explains really quickly?
Something short like "if the pre-computed next card is a buried sibling, discard it since it will be buried" - assuming I understand it correctly

Arthur-Milchior replied:
I rewrite this method, hoping to make code slightly more clear and easy to follow. 

Honestly, I don't like this method, it could be improved quite easily as a single sql request. But I won' change it and keep following upstream right now, as I've not even see it being an important part of the lag in my profiler.

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:
```diff
@@ -135,24 +136,32 @@ public Card getCard() {
         _checkDay();
         if (!mHaveQueues) {
             reset();
+            mNextCard = null;
```
reset() sets mNextCard to null immediately so I think this line is unnecessary?

Arthur-Milchior replied:
Actually, this is theoretically hurtful, since reset asks for a next card in background.

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:
```diff
@@ -2197,6 +2206,10 @@ protected void _burySiblings(Card card) {
                     }
                     mNewQueue.remove(cid);
                 }
+                if (mNextCard != null && mNextCard.getId() == cid && ((buryRev && mNextCard.getQueue() == Consts.QUEUE_TYPE_REV)) || ((buryNew && mNextCard.getQueue() == Consts.QUEUE_TYPE_NEW))) {
```
If I read this correctly it is the exact same as the one in Sched.java, and as mentioned there it is pretty in-depth. Could you make this a method in SchedV2, and then just call it from both? That way  it isn't duplicated

Arthur-Milchior replied:
It's the same line. I was effraid to uselessly add more function and move away even more from upstream.

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:
```diff
@@ -2682,4 +2695,31 @@ public void setReportLimit(int reportLimit) {
 
     /** End #5666 */
+
+    // Not in libanki. In AnkiDroid to save time
```
```suggestion
    // Not in libanki. In AnkiDroid to improve interactive responsiveness
```

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:
```diff
@@ -2682,4 +2695,31 @@ public void setReportLimit(int reportLimit) {
+        _checkDay();
+        if (!mHaveQueues) {
+            reset();
+            mNextCard = null;
```
reset() immediately sets mNextCard = null I think?

mikehardy commented:
Apart from the conflict it looks like it will need some investigation on the connected test too, as 2 of those are failing - 

```
<============-> 99% EXECUTING [1m 2s]> :AnkiDroid:connectedDebugAndroidTest<============-> 99% EXECUTING [1m 3s]<============-> 99% EXECUTING [1m 4s]<============-> 99% EXECUTING [1m 5s]<============-> 99% EXECUTING [1m 6s]<============-> 99% EXECUTING [1m 7s]<============-> 99% EXECUTING [1m 8s]com.ichi2.anki.tests.ContentProviderTest > testSuspendCard[test(AVD) - 4.1.2] FAILED 

        java.lang.NullPointerException

        at com.ichi2.anki.tests.ContentProviderTest.testSuspendCard(ContentProviderTest.java:890)

<============-> 99% EXECUTING [1m 8s]> :AnkiDroid:connectedDebugAndroidTest<============-> 99% EXECUTING [1m 9s]com.ichi2.anki.tests.ContentProviderTest > testAnswerCard[test(AVD) - 4.1.2] FAILED 

        java.lang.RuntimeException: Notes.load(): No result from query for note 1587004418465

        at com.ichi2.libanki.Note.load(Note.java:94)


```

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

Arthur-Milchior commented:
I honestly can't understand what is going on. It works on my phone. in CI, quite a few build failed, but only a minority. Makes no sens

Arthur-Milchior commented:
«FAILURE: Build failed with an exception.» how can the build fail only part of the time ?

mikehardy commented:
That's the normal signature of asynchronous timing issues. It likely means there is some assumption about timing with regard to valid execution state unknowingly built in to the code, and when timing is right, it works, when timing is just slightly off between threads, it does not work.

I haven't reviewed this code so I'm not sure, but when I see that pattern of test results as a developer, I immediately think I have a timing / threading issue and I try to reason through the code with completely different ideas about when what parts execute and if it will be truly safe

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

timrae commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:
```diff
@@ -1572,6 +1577,15 @@ public TaskData doInBackgroundCheckCardSelection(TaskData... params) {
+        try {
+            return new TaskData(lambda.call());
+        } catch (Exception e) {
+            throw new RuntimeException(e);
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293)

timrae replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293)

timrae replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293)

timrae replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293)

Arthur-Milchior commented:
There was one bug I found when I ran the test on my phone a second time. It was related to the way the test are wrote and not to the code itself. In normal use, you can't add a card to the collection and then ask the scheduler for a card without running "reset" at some point.

My code assumed that reset had been called at some point, simply because reset creates empty list and the value are initially "null" (thinking about it, that's probably a mistake in the first place that should be corrected in another PR). So I just asked all test to run "reset" once before getting a card, as would be the case in any real use of AnkiDroid

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

Arthur-Milchior commented:
In this change, I replaced the background task by a more general "Lambda" background task, which execute an arbitrary callable. I realized that for background task which just does one thing and then stops without any action, that's just perfect and could be used in other place too

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

Arthur-Milchior commented:
Yeah, CI solved !!! It seems I was right about the fact that the problem here was that you added card directly in the collection while in real code you can't do this without resetting at least once

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DiscardChangesDialog.java  1
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DiscardChangesDialog.java:
```diff
@@ -0,0 +1,32 @@
+import com.afollestad.materialdialogs.MaterialDialog;
+import com.ichi2.anki.R;
+
+public class DiscardChangesDialog {
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The utility class name 'DiscardChangesDialog' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper)'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293)

timrae replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The utility class name 'DiscardChangesDialog' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper)'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293)

timrae commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:
```diff
@@ -1572,6 +1578,15 @@ public TaskData doInBackgroundCheckCardSelection(TaskData... params) {
+        try {
+            return new TaskData(new Object[]{lambda.call()});
+        } catch (Exception e) {
+            throw new RuntimeException(e);
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293)

timrae replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293)

Arthur-Milchior commented:
Rebased, because I detected a bug with undo and corrected it. 
Undo schedule a reset that should be executed before the next `getCard`. I didn't realize that there was one reason which made important for the reset to occurs later: the current card would be seen as the next card (which makes a lot of sens, the scheduler only knows that reset should occur, not that there is a card currently seen that should be avoided).

So I ensure that the reviewer warns the scheduler that there is a card currently, and so that this card should not be the next card or in the queue.

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
- AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DiscardChangesDialog.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

timrae commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:
```diff
@@ -1573,6 +1578,15 @@ public TaskData doInBackgroundCheckCardSelection(TaskData... params) {
+        try {
+            return new TaskData(new Object[]{lambda.call()});
+        } catch (Exception e) {
+            throw new RuntimeException(e);
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293)

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

timrae commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:
```diff
@@ -59,6 +59,7 @@
 import java.util.Map;
 import java.util.Set;
 import java.util.TreeMap;
+import java.util.concurrent.Callable;
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'java.util.concurrent.Callable'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293)

timrae replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'java.util.concurrent.Callable'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293)

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

Arthur-Milchior commented:
Force push with essentially the same code. The main change was moving a function to some other place in order to ensure that merge with other PR are simpler to do.

Also rebased on master

Arthur-Milchior commented:
The PR is marked as having a requested change. Is there still change you want ? I'd assume that right now the limiting factor is review time and that you want to concentrate on .10 instead of on alpha

mikehardy commented:
I am simply not even re-reviewing them until 2.10alpha is done to preserve focus+attention+time for that effort, so it is possible the request changes mark is stale (so it's signalling value is reduced, sorry). If you have updated the PR *after* I requested changes then it's quite likely it's waiting on me and I haven't re-checked

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

Arthur-Milchior commented:
When you consider this PR, would you please restart API 27 which failed please

Anthropos888 commented:
Short question: does this have an effect on the font loading speed of the webview as well? https://github.com/ankidroid/Anki-Android/issues/5213

Arthur-Milchior commented:
It's literally impossible that it is the cause of any of your current problem. This code is not even in alpha.

In case it's a real question, and not a cause for your problem, then I can tell you that it will not be the case. Fields are fetched from the database, question/answer are computed, but it does not load anything in the webview

Arthur-Milchior commented:
Note that it may actually be a good idea to do it, it's possible to preload data in a browser. But I've not seen it to be a bottle neck currently in my profiler

Arthur-Milchior commented:
Rebased, because there was a merge conflict

Anthropos888 commented:
Sorry, my question was a bit unclear. What I wanted to ask is: might it have a positive effect on the font loading speed if cards are pre-loaded in the background?
I was wondering if the linked issue could be solved by your PR. But as I understand now, your PR does not effect the webviewer. 
Thanks for clarifying.

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

Arthur-Milchior commented:
Rebased to solve merge conflict

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

Arthur-Milchior commented:
same as last time

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

Arthur-Milchior commented:
Rebased to solve a bug found on my phone.
When pre-loading next card, it was removed from count.
Now, the count is not changed until the card is actually returned by getCard

david-allison-1 commented on AnkiDroid/src/androidTest/java/com/ichi2/anki/tests/ContentProviderTest.java:
```diff
@@ -793,6 +793,7 @@ public void testAnswerCard(){
         col = getCol();
         long deckId = mTestDeckIds[0];
         col.getDecks().select(deckId);
+        col.getSched().reset();
```
Agreed that this change is necessary and not a change to the test. Would be useful to "Extract Method" on the 3 lines to reduce the code duplication.

Arthur-Milchior replied:
This has been considered in another PR https://github.com/ankidroid/Anki-Android/pull/6310
Hopefully this file won't be modified in current PR anymore

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -917,6 +918,11 @@ protected void onResume() {
+        // Tells the scheduler there is no more current cards. 0 is
+        // not a valid id.
+        if (mSched != null) {
+            mSched.setCurrentCard(null);
```
`mSched.discardPrefetchedCard` or similar would be more descriptive.

Arthur-Milchior replied:
It is not discarding a "prefetch" card, since this commit/part of the code deal with the current card from the previewer. However, I followed your idea and create a separate method

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:
```diff
@@ -917,6 +918,11 @@ protected void onResume() {
     @Override
     protected void onDestroy() {
         super.onDestroy();
+        // Tells the scheduler there is no more current cards. 0 is
```
I don't see how this occurs. Could you explain this a little more in a PR comment?

Arthur-Milchior replied:
What does "this" refer to ? I assume you're not asking when "onDestroy" is called.

david-allison-1 replied:
>  Tells the scheduler there is no more current cards.

From my reading of the method, it just sets the prefetched card to null

Arthur-Milchior replied:
It sets the current to null. I.e. there is no card currently being reviewed. It does not affect the value mNextCatd

Arthur-Milchior replied:
By the way, can you please confirm that when you states "prefetched" you mean "the next card". Because I'm not entirely sure I understand correctly.

Arthur-Milchior replied:
For information, here is the reason why the Scheduler needs to no the current card of the reviewer.
In current code, the reviewer tells the scheduler to mark the current card as reviewed and then ask for the next card. Since I pre-load the next card before the reviewer's card is marked as reviewed, there is a risk that the selected "next card" is actually a sibling of current card (this is especially the case for new cards), so I need to take care of discarding those siblings. To discard those siblings, I need to know what is the reviewer's card.

Maybe mReviewerCard would be a better variable name than mCurrentCard

david-allison-1 replied:
> By the way, can you please confirm that when you states "prefetched" you mean "the next card". Because I'm not entirely sure I understand correctly.

Yes, the google results aren't too helpful.

> Prefetching in computer science is a technique for speeding up fetch operations by beginning a fetch operation whose result is expected to be needed soon. Usually this is before it is known to be needed ...

https://en.wikipedia.org/wiki/Prefetching

Arthur-Milchior replied:
Do you still have any question unanswered ? I feel like I've not answered "I don't see how this occurs." but I still don't know what "this" refers to.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:
```diff
@@ -1573,6 +1577,12 @@ public TaskData doInBackgroundCheckCardSelection(TaskData... params) {
         return new TaskData(new Object[] { hasUnsuspended, hasUnmarked});
     }
 
+    public TaskData doInBackgroundRunnable(TaskData... params) {
```
I'll flag this up for @mikehardy. I don't like our `TaskData`/`CollectionTask` architecture, so I'm biased, but I don't feel this is going in the right direction.

Arthur-Milchior replied:
Do you know what you want to use instead ?
Since the documentation states that "AsyncTask is designed to be a helper class around Thread and Handler and does not constitute a generic threading framework. AsyncTasks should ideally be used for short operations (a few seconds at the most.)" I should note that I'd be more than happy to see AsyncTask goes. Furthermore, no more than one task run simultaneously in background, which is extremely innefficient since a lot of task could be executed in parallel. In particular the one which read the database and the ones which download. Actually, AsyncTask would not even allow us to download media in background as Anki does since doing saw would block all other background task.

However, if we need to wait for the full task code to be modified in order for this PR to be considered, that's a huge price to pay. The PR to use a correct background system instead of AsyncTask is probably going to be huge and takes time to develop.

david-allison-1 replied:
No, I'm not familiar enough with the Android/Java ecosystem to feel confident suggesting a replacement.

But, this construct worries me:

* There's no error handling specified, it's fire and forget - mostly a problem with our AsyncTask architecture as a whole, but not specifying an error continuation leads to trouble very quickly.
* There's no semantic meaning to "Runnable" - as soon as you begin to allow people to place an arbitrary function into the interface, it becomes much harder to understand the semantics of the calls when migrating to a different system.
* Adding a new field to TaskData:
  * Adds slight memory pressure via the new field, as Java doesn't support unions.
  * Erodes the meaning of TaskData further

Arthur-Milchior replied:
So if I just created a new task number, to represent simply "pre load next card" would it be okay ?
It would still be a "fire and forget" as you state, but I don't see how it is a problem. The very idea of pre-loading data is that I won't need the data immediately but it makes it available for later.

david-allison-1 replied:
Sounds good to me! 

Would also be useful to add a comment stating that exceptions can be thrown without causing state corruption, and do not need to be handled.

Then, a quick manual test to ensure that throwing an exception won't crash the program, and it'll fail gracefully.

Arthur-Milchior replied:
How am I supposed to create such a test ? If I create a backgound task `loadNextCard`, it will simply call the Scheduler method. I do not see how I can cause the methof to fail, except by setting some queue to null or disconnecting/breaking the database.

Arthur-Milchior replied:
How am I supposed to create such a test ? If I create a backgound task `loadNextCard`, it will simply call the Scheduler method. I do not see how I can cause the methof to fail, except by setting some queue to null or disconnecting/breaking the database.

Arthur-Milchior replied:
I've removed Runnable and created a task just for pre-loading the next card. I'll try to make it fail when I know how to do it

david-allison-1 replied:
Manual Test: throw an exception and see how it crashes

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.java:
```diff
@@ -1442,17 +1443,9 @@ protected void _burySiblings(Card card) {
-                    }
-                    // if bury disabled, we still discard to give same-day spacing
-                    mRevQueue.remove(cid);
+                    _actualBurySibling(buryRev, toBury, cid, mRevQueue);
```
It'd be useful to return the value and add it to `toBury` here - removing a parameter from the method

Arthur-Milchior replied:
You want to return which value ?
ids are not always added to toBury, so if you wanted to return the ids to add to toBury it means we would need to put the conditional in the code of _burySiblings and we would have code duplication again.

Furthermore, if I remove the parameter `toBury` then `burying` become useless too

david-allison-1 replied:
On balance, I don't feel the above suggestion went far enough. When I see a mutated collection passed into a method, my gut reaction tells me it's typically much better for readability to extract it and perform the mutation outside the context of the method.

On re-read, I feel this change moves too far away from the style of the Anki code. 

It'd be better to have this as `onBury(cid)` and compartmentalise the additional functionality from the original scheduler logic in a method which is marked as such. 

```java
        // if mNextCard's id is cid, then it should be
        // considered a part of the queue, and so be discarded too.
        if (mNextCard != null && mNextCard.getId() == cid) {
            mNextCard = null;
        }
```

Arthur-Milchior replied:
I am extremely sorry, but I really don't understand your request.
"perform the mutation outside the context of the method", I don't know what is the "context of a method". I can guess that it is not related to the notion of "context" we discussed in another PR. I also assume it's not related to the "scope" in which the method is defined either.

I could, if you want, create a class "CardQueue" which will simply extends List and have a method _actualBury. This way,  it will be clear that we are mutating the queue, and the method doing the mutation is inside the queue class itself. I hope it would solve the problem you have.

I can also simply delete this commit entirely and lead code duplication. It only mean that I'll add ever more code duplication to deal with the current card in both _burySiblings

david-allison-1 replied:
I was referring to the method scope, and the scopes of all of its callees (if any).

----

I think it'd be best to leave the code duplication, as that's present in Anki Desktop. We want to stay as consistent with Anki Desktop as possible.

But, extract the new code to `onBury(long cid)`, so we're not duplicating any additional code.

Arthur-Milchior replied:
When looking at the result, I found that not creating any new method felt more natural. It seemed strange to have a method dealing with burying mNextCard if it is a sibling while duplicating the remaining of the code. So I just entirely removed this function.
I may try to make a better factorization in another PR.

Arthur-Milchior replied:
https://github.com/ankidroid/Anki-Android/pull/6314

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:
```diff
@@ -2186,6 +2246,22 @@ public void buryNote(long nid) {
+        // considered a part of the queue, and so be discarded too.
+        if (mNextCard != null && mNextCard.getId() == cid) {
+            mNextCard = null;
+            CollectionTask.launchCollectionTask(CollectionTask.TASK_TYPE_RUNNABLE,
```
This is Android-specific logic in `libAnki`. Is there a better way?

Arthur-Milchior replied:
Moved. This is now in the Reviewer

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:
```diff
@@ -582,6 +632,12 @@ private void _resetNew() {
         _updateNewCardRatio();
     }
 
+    protected long currentCardNid() {
```
It would be useful to comment here that the return value of 0 is to allow a nicer SQL statement to be created.

Arthur-Milchior replied:
Done, even if I should note that it is not used only in SQL

Arthur-Milchior commented:
> Does this play well with suspend/bury/undo etc..?
Yes. I've been using this code for more than a month, and no problem here. I use undo more than I want to, and I've seen no problem here.

To be more precise, you should note that undo, bury, suspend all call "reset". That is not really efficient, instead of recomputing the queue, we could instead simply take the next element from the queue. However, while it would be far more efficient and still satisfy most of anki's promises (i.e. we would only review cards which need reviewing and which are in the current deck), it means that in some case, the cards are reviewed in a different order than on upstream anki, so it would not be a "NF" change. 
Anyway, my point is that this code plays well with "reset". And since anything different from "review the next card from the queue(filling the queue if required)" trigger a reset, there is no problem here.

Thanks a lot for taking the time to review this PR. I am really looking forward seeing this set of PR processed and seeing ankidroid become efficient on big collections !

Arthur-Milchior commented:
Rebased over master to take into account the tests that are now merged. Most comments of @david-allison-1  taken in consideration.

Arthur-Milchior commented:
Rebased over https://github.com/ankidroid/Anki-Android/pull/6314 since I expect it to be accepted and it simplifies the code

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         

Clones removed
==============
+ AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.java  -1
+ AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

timrae commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:
```diff
@@ -61,6 +61,7 @@
 import java.util.Random;
 
 import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'androidx.annotation.Nullable'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293)

Arthur-Milchior commented:
Rebased over master, since #6314 was accepted.

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5366293&bid=17584382)

Arthur-Milchior commented:
Rebased because of merge conflict

Arthur-Milchior commented:
Merge conflict removed.

Arthur-Milchior commented:
I was going to ask you to restart. And then realized I can now do it myself !!!

Arthur-Milchior commented:
I just realized that there was a off-by-one error on the number of card to review displayed. It displayed I had one more card to review while actually I was reviewing the last card. This is corrected.

Arthur-Milchior commented:
Rebased to use the change from https://github.com/ankidroid/Anki-Android/pull/6404

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:
```diff
@@ -109,6 +109,7 @@
     public static final int TASK_TYPE_SAVE_MODEL = 47;
     public static final int TASK_TYPE_FIND_EMPTY_CARDS = 48;
     public static final int TASK_TYPE_CHECK_CARD_SELECTION = 49;
+    public static final int TASK_TYPE_PRE_LOAD = 50;
```
A very minor thing - can you make the naming more specific?

TASK_TYPE_PRE_LOAD_CARD + doInBackgroundPreLoadCard() ?

Arthur-Milchior replied:
Done

Arthur-Milchior commented:
I'm clearly not asking to put it in main ankidroid without test by a few users. I've been using it for almost two monthes, but experience told me people have different use case than me. Even if to be honest, I can't imagine in which way it'll fail. (But when I had a bug with deck name containing ":::" I had not imagined it either)

mikehardy commented:
Yep - even with complete test coverage there are sometimes surprises. We're still early in 2.12alpha cycle and my hope for it was to get some of these changes in to see how they go with alpha testers. We'll see what @david-allison-1 thinks, I'm ready to merge on it though

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.java:
```diff
@@ -432,8 +432,8 @@ protected boolean _fillLrn() {
                     .query(
-                            "SELECT due, id FROM cards WHERE did IN " + _deckLimit() + " AND queue = " + Consts.QUEUE_TYPE_LRN + " AND due < ? LIMIT ?",
-                                    new Object[]{mDayCutoff, mReportLimit});
+                            "SELECT due, id FROM cards WHERE did IN " + _deckLimit() + " AND queue = " + Consts.QUEUE_TYPE_LRN + " AND due < ? AND nid != ? LIMIT ?",
```
How does this play when "bury siblings" is not checked? Mostly when when we have three cards left in the queue, and they're both from the same note

Arthur-Milchior replied:
I suspect you misunderstand exactly what "bury sibling" does. Please take a look at the comment on https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java#L2231

When you review a card of note N, all cards of note N are removed from the queue. As the comment state: "even if burying disabled, we still discard to give same-day spacing". This is exactly what this line does, it ensure that the queue does not contain sibling of the current card. It does not change the queue type of the card to suspended in any way, and the card will be fetched the next time the queue is filled.

I would gladly admit that this behavior is not documented anywhere, and someone would not know it existed by reading anki(droid)'s manual as far as I know; at least it's not mentioned in https://docs.ankiweb.net/#/studying?id=siblings-and-burying


I may of course add a comment in the code if you want, so that the next reader does not wonder about "bury sibling".


If you want to check in details that my PR does what it should do, you can try to follow along the method "answer card". My PR tries to simulate answerCard's behaviour, while not actually changing anything about the current card. This is more complex than it sounds, because that method is quite complex. However, a lot of complexity deals with computing the next interval, the side effect on the queues and the counts are really limited.

Arthur-Milchior replied:
I put the thumb for "please add comment", so I did.
Seems useful even if I misunderstood your thumb

Arthur-Milchior commented:
Push force to remove merge conflict

mikehardy commented:
@david-allison-1 this could merge if you approve

david-allison-1 commented:
@mikehardy I'll aim to get this done by tonight. I'm still not sure that I understand the implications

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:
```diff
@@ -105,6 +106,25 @@
+    /**
+     * The card currently being reviewed.
+     *
+     * Should be skipped is found to be the next card. This allow to
```
Consider condensing both lines to:
```suggestion
     * Must not be returned during prefetching (as it is currently shown).
```

Arthur-Milchior replied:
Done

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:
```diff
@@ -105,6 +106,25 @@
+
+        Null iff mNextCard is null.*/
+    @Nullable
+    protected List<Long> mCurrentCardParents;
```
```suggestion
    protected List<Long> mCurrentCardParentDids;
```

Arthur-Milchior replied:
Done

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:
```diff
@@ -105,6 +106,25 @@
+     * look for next card during a reset while a card is in reviewer.
+     */
+    protected Card mCurrentCard;
+    /** The list of parent of current card. We save it as a member
```
```suggestion
    /** The list of parent decks of the current card. Cached for performance 
    ...
    */
```

Arthur-Milchior replied:
Done

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:
```diff
@@ -2718,4 +2826,34 @@ public void setReportLimit(int reportLimit) {
     /** End #5666 */
+    public void discardCurrentCard() {
+        mCurrentCard = null;
+        mCurrentCardParents = null;
```
Can this cause `NullPointerExceptions` on concurrent accesses?

Arthur-Milchior replied:
I test that the value is non null before each access. However, I believe you're right, if you close the reviewer (and thus call `discardCurrentCard`) while some background scheduler task occurs, and `mCurrentCard` is set to null between the test to null and the access to the value, there will be a null pointer exception. I'll correct that, thanks

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:
```diff
@@ -2718,4 +2826,34 @@ public void setReportLimit(int reportLimit) {
+    }
+
+    /**
+     * This imitate the action of the method answerCard, except that
```
```suggestion
     * This imitates the action of the method answerCard, except that
```

Arthur-Milchior replied:
Done

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:
```diff
@@ -2718,4 +2826,34 @@ public void setReportLimit(int reportLimit) {
+
+    /**
+     * This imitate the action of the method answerCard, except that
+     * it does not state of any card. It means in particular that:
```
"it does not state of" - what's the meaning?

Arthur-Milchior replied:
"it dose not change the state of"... sorry

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:
```diff
@@ -2718,4 +2826,34 @@ public void setReportLimit(int reportLimit) {
+        // if current card is next card or in the queue
+        if (mNextCard != null && mCurrentCard.getId() == mNextCard.getId()) {
+            mNextCard = null;
+            CollectionTask.launchCollectionTask(CollectionTask.TASK_TYPE_PRE_LOAD_CARD);
```
I feel I've mentioned this before, but what are the other architectures that we could use to avoid adding `AsyncTask` into `libAnki`?

Arthur-Milchior replied:
We already discussed similar topic. In an older version of this PR, I was using a collection task which was simply "call this method in background". I.e. the task takes a runnable and run it. You thought it was a big risk and instead you wanted that the card type simply does one thing. This is why I know have a type "pre_load_card". 

We also mentioned that AsyncTask is really not a good choice and should be replaced by a system which is not deprecated, able to handle multiple background tasks simultaneously, but that it would be an extremely big change and probably breaks a lot of thing. So that's not something I'll do here.

I do not recall discussing what could be done instead of async task in libanki specifically. Honestly, I've no idea what could be done instead. In particular, I believe using a background action is correct here, it is an heuristict to make the app quicker, and not an action mandatory for the correctness of the method, so sending this action is background makes sens here.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:
```diff
@@ -149,26 +169,68 @@ public Card getCard() {
     }
 
     public void reset() {
+        mNextCard = null;
```
This doesn't unset the parents? Intentional?

Arthur-Milchior replied:
No. Thank you.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/AbstractSched.java:
```diff
@@ -263,4 +266,19 @@ public void run() {
+    /**
+       set mNextCard to the next card that should be reviewed. Fill
+       queue if required.
+       @param qa Whether to also pre-computer question and answer
```
Corrected

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/AbstractSched.java:
```diff
@@ -263,4 +266,19 @@ public void run() {
+    public abstract Card loadNextCard(boolean qa);
+
+    /**
+     * Indicates the scheduler that a card is currently being reviewed, so it should not be preparred as the next card.
```
Currently, the scheduler does not know which card is being reviewed. Usually, the last card returned by `getCard` is getting reviewed. However, often, it is not the case.
First, if you leave the reviewer, the scheduler does not get the information. It may eventually gets the information that it needs to "reset" and so but the current card to null. However, if you leave the deck and go back to the same deck immediately, "reset" is not necessarily called, and so the scheduler will believe the reviewer has a card, while actually it is not the case anymore.
The reverse problem may occur, the reviewer does not necessarily get its cards from the scheduler. I believe the only other way to get a card is "undo". Of course, I could also asks "undo" to warns the scheduler that it sends a card back to the reviewer; but I believe it would makes things far more complex than just letting the reviewer notifies the scheduler.
To be more precise, if you review a card `C` and undo the review, a message is sent to the scheduler that it should "reset". The reviewer gets card `C` immediately, so it can be shown quickly to the user. However, when the scheduler prefetch the next card, it will usually believes that the next card is card `C`, and so it will prefetch it. This is not a problem in itself, because when `C` gets reviewed again, the scheduler will discard the prefetched card, but it is time lost. Since I let the scheduler knows that `C` is currently getting reviewed, the scheduler can prefetch the correct next card.


I don't know whether those explanation makes things more clear. I hope they do. If it is the case, I would love help to create meaningful and simple comments, because I've little idea how to short commets that explain all of this

Arthur-Milchior commented:
I can hardly imagine how you'd get to see the same card twice, except if there is no card in learning between them. If I read `_answerLrnCard` and `_rescheduleLrnCard` correctly, those methods ensure that as soon as there are two cards in learning, you alternate between them, and you never see the same card again successively.

If there is a single card in learning, you'll see the same card twice, indeed.

In any case, you are entirely right, the current card will never be prefetched. However, you can see the same card twice, however this only occurs if there is no other card to review now. In this case, the scheduler will works as follow: it'll realize that no card were pre-fetched, and so try to fetch a card as it was doing before this PR, and it'll get the card you just reviewed back.

david-allison-1 commented:
Firstly, I'm really sorry it's taken this long for me to get back around to it.

At the current moment, I'm not confident that I understand this change well enough to maintain it if anything goes wrong. I understand what it's meant to do, but I've still been misunderstanding the implementation in this review. That's a me problem, but it makes me hesitate. I'll need to give it another look with a clear mind tomorrow.

Please get in contact if you don't hear back within 24 hours.

Arthur-Milchior commented:
I mean, if I can't explain why this work convincigly, this is clearly a part of the problem.

I wonder whether a part of the problem is not that the scheduler itself is complex. You state that you don't understand my change, may I ask if you understand the scheduler in the first place ? I hope this question is not insulting, I can tell you it took me a lot of time to understand it myself and write the documentation in https://github.com/Arthur-Milchior/anki/blob/commented/documentation/computing_intervals.md . So, I would expect that if you had no reason to deal with the scheduler yourself, you have little idea how it actually works. And if you don't know how it works, I expect it to be hard to understand the change made to it.

Which would actually be a bigger problem than just this PR. If there were change made in the scheduler before we incorporate rust in ankidroid, it's going to be quite hard to port them to ankidroid

Arthur-Milchior commented:
If you want help with the scheduler, I just wrote a document trying to explain how it works: https://github.com/Arthur-Milchior/anki/blob/commented/documentation/scheduler.md

Arthur-Milchior commented:
@david-allison-1 : this is your 24 hours reminder.

Arthur-Milchior commented:
Removed merge conflict

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/sched/SchedV2.java:
```diff
@@ -148,7 +168,8 @@ public Card getCard() {
             reset();
         }
-        Card card = _getCard();
+        Card card = loadNextCard(false);
```
Flagging that this will potentially change the returned value.

`_getLrnCard()` depends on the current time, by performing this operation early, a learn card can be skipped.

Arthur-Milchior replied:
I admit I had not thought about this one. You're entirely right. I could either:
* check for card in learning before returning the pre fetched next card
* state that if two cards are inversed, it's an acceptable change in order to win some time. Assuming a card is reviewed in 30 seconds, the card in learning will be delayed by 30 seconds max. In term of learning, it should not make a big difference. If the user spend an hour to review a card, then there is probably some other problem than this change of order

david-allison-1 replied:
Best bet would be to check that we're returning the correct card, the further we move from Anki Desktop, the harder it is to test.

Arthur-Milchior commented:
Conflict removed

Arthur-Milchior commented:
Rebased. Now that the current card is known by the scheduler, a single commit remain

Arthur-Milchior commented:
Merge conflict removed

mikehardy commented:
@david-allison-1 - no pressure of course, but you had some comments I think were addressed

Arthur-Milchior commented:
Moving to draft right now. I had two related ideas thanks to David's comment. No time to code it yet

I don't need to pre fetch the next card if it is a card in learning from today. It is necessary in the queue, so I'll just pre-fetch the next card not in learning.

I'll also move CardCache outside of card browser. Having a data structure which contains an id and potentially the card seems perfect here. I can leave the card in the queue, which will simplify the change, and find the head of the queue to tell them to load and pre-compute question answer

Arthur-Milchior commented:
A simpler verison was added on https://github.com/ankidroid/Anki-Android/pull/6850
